Panic / segfault when using query command with --arch option that doesn't match the platform

It appears that, starting with Please v17.4.0, using the query command with the option `--arch=windows_amd64` while running on `linux_amd64` results in a segfault:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x278 pc=0x872bea]

goroutine 50 [running]:
github.com/thought-machine/please/src/core.(*BuildTarget).State(...)
        src/core/build_target.go:1086
github.com/thought-machine/please/src/core.(*BuildState).waitForTargetAndEnsureDownload(0xc0000ca000?, {{0x0, 0x0}, {0xc000419690, 0xc}, {0x0, 0x0}}, {{0x0, 0x0}, {0xe61aa0, ...}, ...}, ...)
        src/core/state.go:983 +0x8a
github.com/thought-machine/please/src/core.(*BuildState).WaitForInitialTargetAndEnsureDownload(...)
        src/core/state.go:978
github.com/thought-machine/please/src/plz.findOriginalTask(0xc0000ca000, {{0x0, 0x0}, {0xe61aa0, 0x3}, {0xc000419690, 0xc}}, 0x1, {{0x7ffd483785a4, 0x6}, ...})
        src/plz/plz.go:170 +0x625
github.com/thought-machine/please/src/plz.findOriginalTaskSet(0x0?, {0x1556a80?, 0x6?, 0x7ffd483785ab?}, 0x5?, {{0x7ffd483785a4?, 0x0?}, {0x7ffd483785ab?, 0x0?}})
        src/plz/plz.go:134 +0x119
github.com/thought-machine/please/src/plz.findOriginalTasks(0xc0000ca000, {0x0?, 0x0, 0x0}, {0x1556a80?, 0x1, 0x1}, {{0x7ffd483785a4, 0x6}, {0x7ffd483785ab, ...}})
        src/plz/plz.go:127 +0x3e9
created by github.com/thought-machine/please/src/plz.Run in goroutine 1
        src/plz/plz.go:40 +0x3b2
```

These fail:

`plz query alltargets --arch=darwin_amd64`
`plz query alltargets --arch=windows_amd64`

But this succeeds:

`plz query alltargets --arch=linux_amd64`

The original reason I found this was because we have an alias to run a custom rule that handles uploading releases to our internal distribution system, which boiled down to something like:

`plz run --arch=windows_amd64 sequential --include=dist ...`

After doing some debugging, I was able to consistently reproduce using the above `query` command.

I am trying to put together a simple please project I can use to reproduce these kinds of errors to share back with the team, but it's not yet done and I wanted to get the issue in quickly in case there was something more high-level that I'm doing wrong.

--- a/src/core/state.go
+++ b/src/core/state.go
@@ -975,6 +975,10 @@ func (state *BuildState) WaitForTargetAndEnsureDownload(l, dependent BuildLabel,
 // WaitForInitialTargetAndEnsureDownload is like WaitForTargetAndEnsureDownload but is used for
 // targets in the initial set.
 func (state *BuildState) WaitForInitialTargetAndEnsureDownload(l, dependent BuildLabel) *BuildTarget {
+	// This may have been an architecture label from the CLI
+	if state.WaitForBuiltTarget(l, dependent, ParseModeNormal) == nil {
+		return nil
+	}
 	return state.waitForTargetAndEnsureDownload(l, dependent, ParseModeNormal)
 }
 

--- a/src/plz/plz.go
+++ b/src/plz/plz.go
@@ -165,14 +165,15 @@ func findOriginalTask(state *core.BuildState, target core.BuildLabel, addToList 
 		// walk the directory tree, so we have to make sure the subrepo exists first.
 		dir := target.PackageName
 		prefix := ""
-		if target.Subrepo != "" && target.Subrepo != arch.String() {
+		if target.Subrepo != "" {
 			subrepoLabel := target.SubrepoLabel(state)
-			state.WaitForInitialTargetAndEnsureDownload(subrepoLabel, target)
-			// Targets now get activated during parsing, so can be built before we finish parsing their package.
-			state.WaitForPackage(subrepoLabel, target, core.ParseModeNormal)
-			subrepo := state.Graph.SubrepoOrDie(target.Subrepo)
-			dir = subrepo.Dir(dir)
-			prefix = subrepo.Dir(prefix)
+			if state.WaitForInitialTargetAndEnsureDownload(subrepoLabel, target) != nil {
+				// Targets now get activated during parsing, so can be built before we finish parsing their package.
+				state.WaitForPackage(subrepoLabel, target, core.ParseModeNormal)
+				subrepo := state.Graph.SubrepoOrDie(target.Subrepo)
+				dir = subrepo.Dir(dir)
+				prefix = subrepo.Dir(prefix)
+			}
 		}
 		for filename := range FindAllBuildFiles(state.Config, dir, "") {
 			dirname, _ := filepath.Split(filename)

--- a/src/plz/plz.go
+++ b/src/plz/plz.go
@@ -165,7 +165,7 @@ func findOriginalTask(state *core.BuildState, target core.BuildLabel, addToList 
 		// walk the directory tree, so we have to make sure the subrepo exists first.
 		dir := target.PackageName
 		prefix := ""
-		if target.Subrepo != "" {
+		if target.Subrepo != "" && target.Subrepo != arch.String() {
 			subrepoLabel := target.SubrepoLabel(state)
 			state.WaitForInitialTargetAndEnsureDownload(subrepoLabel, target)
 			// Targets now get activated during parsing, so can be built before we finish parsing their package.

